File: 13. Next Greater Element - GFG/13.-next-greater-element.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution
{
    public:
    //Function to find the next greater element for each element of the array.
    vector<long long> nextLargerElement(vector<long long> arr, int n){
        vector <long long> v;
        stack<long long> s;
        s.push(arr[n-1]);
        long long x;
        v.push_back(-1);
        for(int i = n-2; i >= 0; i--)
        {
            while(!s.empty() && arr[i] >= s.top())           
                s.pop();
            x = (s.empty())? -1 : s.top();
            v.push_back(x);
            s.push(arr[i]);
        }
        reverse(v.begin(), v.end());
        return v;
    }

};

// { Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        
        int n;
        cin>>n;
        vector<long long> arr(n);
        for(int i=0;i<n;i++)
            cin>>arr[i];
        
        Solution obj;
        vector <long long> res = obj.nextLargerElement(arr, n);
        for (long long i : res) cout << i << " ";
        cout<<endl;
    }
	return 0;
}  // } Driver Code Ends

File: 13. Next Greater Element - GFG/13.-next-greater-element.py
#User function Template for python3


class Solution:
    def nextLargerElement(self,arr,n):
        l = []
        stack = []
        l.append(-1)
        stack.append(arr[-1])
        for i in range(n-2,-1,-1):
            while(len(stack)>0 and arr[i] >= stack[-1]): stack.pop()
            
            if(len(stack) == 0): x = -1
            else: x = stack[-1]
            l.append(x)
            stack.append(arr[i])
        
        l = l[::-1]
        return l

#{ 
#  Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases) :
        n = int(input())

        a = list(map(int,input().strip().split()))
        obj = Solution()
        res = obj.nextLargerElement(a,n)
        for i in range (len (res)):
            print (res[i], end = " ")
        print ()
# } Driver Code Ends